bpf/lib/maps.h

```diff
@@ -77,6 +78,180 @@ struct bpf_elf_map __section_maps CALLS_MAP = {
 	.max_elem	= CILIUM_CALL_SIZE,
 };
 
+#ifdef POLICY_ENFORCEMENT
+
+#ifdef HAVE_LPM_MAP_TYPE
+struct bpf_lpm_trie_key6 {
+	struct bpf_lpm_trie_key lpm_key;
+	union v6addr lpm_addr;
+};
+
+static __always_inline int lpm6_map_lookup(struct bpf_elf_map *map, union v6addr *addr)
+{
+	struct bpf_lpm_trie_key6 key = { { 128 }, *addr };
+	return map_lookup_elem(map, &key) != NULL;
+}
+
+#ifdef CIDR6_INGRESS_MAP
+struct bpf_elf_map __section_maps CIDR6_INGRESS_MAP = {
+	.type		= BPF_MAP_TYPE_LPM_TRIE,
+	.size_key	= sizeof(struct bpf_lpm_trie_key6),
+	.size_value	= sizeof(struct policy_entry),
+	.pinning	= PIN_GLOBAL_NS,
+	.max_elem	= 1024,
+	.flags		= BPF_F_NO_PREALLOC,
+};
+#define lpm6_ingress_lookup(ADDR) lpm6_map_lookup(&CIDR6_INGRESS_MAP, ADDR)
+#else
+#define lpm6_ingress_lookup(ADDR) 0
+#endif
+
+#ifdef CIDR6_EGRESS_MAP
+struct bpf_elf_map __section_maps CIDR6_EGRESS_MAP = {
+	.type		= BPF_MAP_TYPE_LPM_TRIE,
+	.size_key	= sizeof(struct bpf_lpm_trie_key6),
+	.size_value	= sizeof(struct policy_entry),
+	.pinning	= PIN_GLOBAL_NS,
+	.max_elem	= 1024,
+	.flags		= BPF_F_NO_PREALLOC,
+};
+#define lpm6_egress_lookup(ADDR) lpm6_map_lookup(&CIDR6_EGRESS_MAP, ADDR)
+#else
+#define lpm6_egress_lookup(ADDR) 0
+#endif
+
+struct bpf_lpm_trie_key4 {
+	struct bpf_lpm_trie_key lpm_key;
+	__be32 lpm_addr;
+};
+
+static __always_inline int lpm4_map_lookup(struct bpf_elf_map *map, __be32 addr)
+{
+	struct bpf_lpm_trie_key6 key = { { 32 }, addr };
+	return map_lookup_elem(map, &key) != NULL;
+}
+
+#ifdef CIDR4_INGRESS_MAP
+struct bpf_elf_map __section_maps CIDR4_INGRESS_MAP = {
+	.type		= BPF_MAP_TYPE_LPM_TRIE,
+	.size_key	= sizeof(struct bpf_lpm_trie_key4),
+	.size_value	= sizeof(struct policy_entry),
+	.pinning	= PIN_GLOBAL_NS,
+	.max_elem	= 1024,
+	.flags		= BPF_F_NO_PREALLOC,
+};
+#define lpm4_ingress_lookup(ADDR) lpm4_map_lookup(&CIDR4_INGRESS_MAP, ADDR)
+#else
+#define lpm4_ingress_lookup(ADDR) 0
+#endif
+#ifdef CIDR4_EGRESS_MAP
+struct bpf_elf_map __section_maps CIDR4_EGRESS_MAP = {
+	.type		= BPF_MAP_TYPE_LPM_TRIE,
+	.size_key	= sizeof(struct bpf_lpm_trie_key4),
+	.size_value	= sizeof(struct policy_entry),
+	.pinning	= PIN_GLOBAL_NS,
+	.max_elem	= 1024,
+	.flags		= BPF_F_NO_PREALLOC,
+};
+#define lpm4_egress_lookup(ADDR) lpm4_map_lookup(&CIDR4_EGRESS_MAP, ADDR)
+#else
+#define lpm4_egress_lookup(ADDR) 0
+#endif
+
+#else /* HAVE_LPM_MAP_TYPE */
+/* No LPM map, use an array instead. Since our policies are default
+ * deny we can stop at the first match. */
+
+struct cidr6_entry {
+	union v6addr net, mask;
+};
+
+#ifdef CIDR6_INGRESS_MAPPINGS
+static __always_inline int lpm6_ingress_lookup(union v6addr *addr)
+{
+	struct cidr6_entry map[] = { CIDR6_INGRESS_MAPPINGS };
+	const int size = (sizeof(map) / sizeof(map[0]));
+	int i;
+
+#pragma unroll
```

Comment:
Seem to ge working for now, but will have to see when this code is exposed to more testing in different kernels, llvm versions, etc.